uqcsbot/utils/snailrace_utils.py

```diff
@@ -0,0 +1,146 @@
+SNAILRACE_ENTRY_CLOSE = "Entry is now closed. Let's Race!"
+SNAILRACE_JOIN = "%s has Joined the Race!"
+SNAILRACE_ALREADY_JOINED = "%s is already apart of the race."
+SNAILRACE_BOARD = "And they're off...\n```\n%s\n```"
+SNAILRACE_WINNER = "The race has finished! %s has won!"
+SNAILRACE_NO_START = "Sorry, but there aren't enough racers to start the race!"
+
+class SnailRacer:
+    def __init__(self, member: discord.Member, racer_id: int):
+        self.member = member
+        self.racer_id = racer_id
+
+        self.position = 0
+        self.step_number = 0
+    
+    def step(self):
+        # Calculate the number of positions the snail will take this step
+        speed = random.randint(SNAILRACE_MIN_STEP, SNAILRACE_MAX_STEP)
+        self.position = min(self.position + speed, SNAILRACE_TRACK_LENGTH)
+
+        # If the snail has reached the end of the track, it can't move anymore
+        if not self.position >= SNAILRACE_TRACK_LENGTH:
+            self.step_number += 1
+
+    def __str__(self):
+        # Calculate the index of the snail emoji in the track
+        index = min(
+            int((self.position / SNAILRACE_TRACK_LENGTH) * SNAILRACE_TRACK_LENGTH_CHARS), 
+            SNAILRACE_TRACK_LENGTH_CHARS - 1
+        )
+
+        # Create the track string
+        track = SNAILRACE_TRACK_LENGTH_CHARS * [" "]
+        track[:index] = ["."] * index
+        track[index] = SNAILRACE_SNAIL_EMOJI
+        track_str = "".join(track)
+
+        return f"{str(self.racer_id).rjust(3)} |{track_str}| {self.member.display_name}"
+
+
+class SnailRaceState:
+    def __init__(self):
+        self.race_start_time = datetime.datetime.now()
+        self.racing = False
+        
+        self.racers = []
+        self.open_interaction = None
+
+    def is_racing(self) -> bool:
+
+        # Calculate max race time with a 5 second bias
+        max_race_len = SNAILRACE_OPEN_TIME + (SNAILRACE_STEP_TIME * SNAILRACE_TRACK_LENGTH) + 5
+        if (datetime.datetime.now() - self.race_start_time).total_seconds() >= max_race_len:
+            self.close_race()
+
+        return self.racing
+
+    def open_race(self, open_interaction: discord.Interaction):
+        # Start the entry
+        self.racing = True
+        self.open_interaction = open_interaction
+
+    def close_race(self):
+        # Close the race for a new one to start
+        self.racing = False
+        self.racers = []
+        self.open_interaction = None
+
+    def add_racer(self, racer: discord.Member) -> bool:
+        # Filter Unique Racers
+        for r in self.racers:
+            if r.member.id == racer.id:
+                return False
+
+        self.racers.append(SnailRacer(racer, len(self.racers) + 1))
+        return True
+    
+    async def race_start(self):
+        await self._start_racing(self.open_interaction)
+
+    async def _start_racing(self, interaction: discord.Interaction):
```

Comment:
General comment:

I like this much better than passing around "function pointers" but it does kinda change the purpose of splitting up `SnailRace` with IO and `SnailRaceState` with game logic.

I won't call this a blocking issue because then `SnailRace` just has to do the interaction handler, setup, & initial command, and if you get the desire to add a leaderboard (as previously mentioned) or other commands, it's nice to have `SnailRace` as the general "I've received a request" handler with other files/functions for actually processing these requests. So, carry on, I guess, just a note explaining my review thought process.

Reply:
Yeah thats fair, It will probably need to be revisited when the `snailrace` gets further development later on. But I'll keep this in mind.